src/ui/edit_field.c

```diff
@@ -20,14 +20,270 @@ struct Edit_field
+// For an explanation of the naming terminology for these helper methods
+static bool is_emacs_word(char c);
+static void forward_char(Edit_field *edit_field);
+static void backward_char(Edit_field *edit_field);
+static void move_beginning_of_line(Edit_field *edit_field);
+static void move_end_of_line(Edit_field *edit_field);
+static void forward_word(Edit_field *edit_field);
+static void backward_word(Edit_field *edit_field);
+static void delete_char(Edit_field *edit_field);
+static void delete_backward_char(Edit_field *edit_field);
+static void kill_word(Edit_field *edit_field);
+static void backward_kill_word(Edit_field *edit_field);
+
+static void handle_keydown(Edit_field *edit_field, const SDL_Event *event);
+static void handle_keydown_alt(Edit_field *edit_field, const SDL_Event *event);
+static void handle_keydown_ctrl(Edit_field *edit_field, const SDL_Event *event);
+
+static void edit_field_insert_char(Edit_field *edit_field, char c)
+{
+    if (edit_field->buffer_size >= BUFFER_CAPACITY) {
+        return;
+    }
+
+    for (int64_t i = (int64_t) edit_field->buffer_size - 1; i >= (int64_t) edit_field->cursor; --i) {
+        edit_field->buffer[i + 1] = edit_field->buffer[i];
+    }
+
+    edit_field->buffer[edit_field->cursor++] = c;
+    edit_field->buffer[++edit_field->buffer_size] = 0;
+}
+
+// See: https://www.gnu.org/software/emacs/manual/html_node/emacs/Moving-Point.html
+// For an explanation of the naming terminology for these helper methods
+
+static bool is_emacs_word(char c)
+{
+#define IN_RANGE(start, end) (c >= start && c <= end)
+
+    // Word syntax table retrieved from Fundamental Mode, "C-h s"
+    // (This is not the complete syntax table)
+    return IN_RANGE('$', '%')
+        || IN_RANGE('0', '9')
+        || IN_RANGE('A', 'Z')
+        || IN_RANGE('a', 'z')
+        ;
+
+#undef IN_RANGE
+}
+
+static void forward_char(Edit_field *edit_field)
+{
+    // "C-f" or "<RIGHT>"
+    if (edit_field->cursor < edit_field->buffer_size) {
+        edit_field->cursor++;
+    }
+}
+
+static void backward_char(Edit_field *edit_field)
+{
+    // "C-b" or "<LEFT>"
+    if (edit_field->cursor > 0) {
+        edit_field->cursor--;
+    }
+}
+
+static void move_beginning_of_line(Edit_field *edit_field)
+{
+    // "C-a" or "<Home>"
+    edit_field->cursor = 0;
+}
+
+static void move_end_of_line(Edit_field *edit_field)
+{
+    // "C-e" or "<End>"
+    edit_field->cursor = edit_field->buffer_size;
+}
+
+static void forward_word(Edit_field *edit_field)
+{
+    // "M-f" or "C-<RIGHT>" or "M-<RIGHT>"
+    while (true) {
+        forward_char(edit_field);
+        if (edit_field->cursor >= edit_field->buffer_size) {
+            break;
+        }
+
+        char current = edit_field->buffer[edit_field->cursor];
+        char preceeding = edit_field->buffer[edit_field->cursor - 1];
+        if (!is_emacs_word(current) && is_emacs_word(preceeding)) {
+            // Reached the end of the current word
+            break;
+        }
+    }
```

Comment:
Why can't this be something like `while(cursor < buffer_size && is_emacs_word(buffer[cursor])) { ++cursor; }`?  Why does it have to be so over-complicated?

Reply:
I agree that it seems overcomplicated, but that really is the necessary logic for jumping to the end of the next word. You must:

1. Move forward one character, no matter whether the current character is a word or not
2. Keep moving forward until the current character no longer counts as a "word", *and* the previous character counted as a word.

It's not sufficient to stop once the current character is no longer a word because of the case where there's multiple "non-word" characters in a row.

This is an alternative:

```c
    do {
        forward_char(edit_field);
    } while (edit_field->cursor < edit_field->buffer_size
             && (is_emacs_word(edit_field->buffer[edit_field->cursor])
                 || !is_emacs_word(edit_field->buffer[edit_field->cursor - 1])));
```

Which would you prefer?

Reply:
Ok, I see. Thank you for such a detailed explanation. I'll take a closer look into this PR tomorrow (after some sleep). Thank you for your time and contribution!